src/python/grpcio_testing/grpc_testing/_time.py

```diff
@@ -0,0 +1,223 @@
+import threading
+import time as _time
+
+import grpc
+import grpc_testing
+
+
+def _call(behaviors):
+    for behavior in behaviors:
+        try:
+            behavior()
+        except Exception:  # pylint: disable=broad-except
+            logging.exception('Exception calling behavior?')
+
+
+def _call_in_thread(behaviors):
+    calling = threading.Thread(target=_call, args=(behaviors,))
+    calling.start()
+    calling.join()
+
+
+class _State(object):
+
+    def __init__(self):
+        self.condition = threading.Condition()
+        self.times_to_behaviors = collections.defaultdict(list)
+        self.behaviors_to_times = {}
+
+
+class _Delta(
+        collections.namedtuple('_Delta',
+                               ('mature_behaviors', 'earliest_mature_time',
+                                'earliest_immature_time',))):
+    pass
+
+
+def _process(state, now):
+    mature_behaviors = []
+    earliest_mature_time = None
+    while state.times_to_behaviors:
+        earliest_time = min(state.times_to_behaviors)
+        if earliest_time <= now:
+            if earliest_mature_time is None:
+                earliest_mature_time = earliest_time
+            earliest_mature_behaviors = state.times_to_behaviors.pop(
+                earliest_time)
+            for mature_behavior in earliest_mature_behaviors:
+                state.behaviors_to_times.pop(mature_behavior)
+            mature_behaviors.extend(earliest_mature_behaviors)
+        else:
+            earliest_immature_time = earliest_time
+            break
+    else:
+        earliest_immature_time = None
+    return _Delta(mature_behaviors, earliest_mature_time,
+                  earliest_immature_time)
+
+
+class _Future(grpc.Future):
+
+    def __init__(self, state, behavior):
+        self._state = state
+        self._behavior = behavior
+        self._cancelled = False
+
+    def cancel(self):
+        with self._state.condition:
+            time = self._state.behaviors_to_times.pop(self._behavior, None)
+            if time is None:
+                return False
+            else:
+                self._state.times_to_behaviors[time].remove(self._behavior)
+                if not self._state.times_to_behaviors[time]:
+                    self._state.times_to_behaviors.pop(time)
+                    self._state.condition.notify_all()
+                self._cancelled = True
+                return True
+
+    def cancelled(self):
+        with self._state.condition:
+            return self._cancelled
+
+    def running(self):
+        raise NotImplementedError()
+
+    def done(self):
+        raise NotImplementedError()
+
+    def result(self, timeout=None):
+        raise NotImplementedError()
+
+    def exception(self, timeout=None):
+        raise NotImplementedError()
+
+    def traceback(self, timeout=None):
+        raise NotImplementedError()
+
+    def add_done_callback(self, fn):
+        raise NotImplementedError()
+
+
+class RealTime(grpc_testing.Time):
```

Comment:
Actually, the way your doing it is pretty close to what I described, but using a thread instead of an interrupt.

If I were implementing it, I'd have the _activity() thread sleep on a threading.Event() with a timeout. The timeout would be set for the next call time, with the option to wake earlier if a call to _call_at() came in earlier. But I think that's a coin toss.